File: android/src/main/kotlin/com/tsitser/background_http_plugin/data/datasource/TaskQueueManager.kt
package com.tsitser.background_http_plugin.data.datasource

import android.content.Context
import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.json.JSONArray
import java.io.File
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.CopyOnWriteArraySet

/**
 * Task queue manager for controlling the number of concurrent requests.
 *
 * Solves the problem of hangs when registering a large number of tasks in WorkManager.
 * Instead of registering all tasks in WorkManager at once,
 * TaskQueueManager keeps them in its own queue and starts them in batches.
 */
class TaskQueueManager(private val context: Context) {
    
    companion object {
        private const val TAG = "TaskQueueManager"
        private const val QUEUE_FILE_NAME = "pending_queue.json"
        private const val DEFAULT_MAX_CONCURRENT_TASKS = 30
        private const val DEFAULT_MAX_QUEUE_SIZE = 10000
        
        @Volatile
        private var instance: TaskQueueManager? = null
        
        fun getInstance(context: Context): TaskQueueManager {
            return instance ?: synchronized(this) {
                instance ?: TaskQueueManager(context.applicationContext).also { instance = it }
            }
        }
    }
    
    private val workManager = WorkManagerDataSource(context)
    
    // Queue of pending tasks (only requestId, data is on disk)
    private val pendingQueue = ConcurrentLinkedQueue<String>()
    
    // Set of active tasks (started in WorkManager)
    private val activeTasks = CopyOnWriteArraySet<String>()
    
    // Mutex for synchronizing queue operations
    private val queueMutex = Mutex()
    
    // Coroutine scope for background operations
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    // Settings
    @Volatile
    var maxConcurrentTasks: Int = DEFAULT_MAX_CONCURRENT_TASKS
        private set
    
    @Volatile
    var maxQueueSize: Int = DEFAULT_MAX_QUEUE_SIZE
        private set
    
    // File for persistent queue storage
    private val queueFile: File
        get() {
            val dir = File(context.filesDir, "background_http_client")
            if (!dir.exists()) dir.mkdirs()
            return File(dir, QUEUE_FILE_NAME)
        }
    
    init {
        // Restore queue on initialization
        restoreQueue()
    }
    
    /**
     * Adds a task to the queue.
     * @param requestId task ID
     * @return true if the task was added, false if the queue is full
     */
    suspend fun enqueue(requestId: String): Boolean = queueMutex.withLock {
        // Check that the queue size limit is not exceeded
        if (pendingQueue.size >= maxQueueSize) {
            Log.w(TAG, "Queue is full ($maxQueueSize), rejecting task: $requestId")
            return@withLock false
        }
        
        // Check that this task has not already been added
        if (pendingQueue.contains(requestId) || activeTasks.contains(requestId)) {
            Log.d(TAG, "Task already in queue or active: $requestId")
            return@withLock true
        }
        
        pendingQueue.offer(requestId)
        saveQueueAsync()
        
        Log.d(TAG, "Task enqueued: $requestId, queue size: ${pendingQueue.size}, active: ${activeTasks.size}")
        
        // Try to start tasks
        processQueueInternal()
        
        return@withLock true
    }
    
    /**
     * Called when a task is completed (success or error).
     */
    suspend fun onTaskCompleted(requestId: String) = queueMutex.withLock {
        if (activeTasks.remove(requestId)) {
            Log.d(TAG, "Task completed: $requestId, active: ${activeTasks.size}")
            processQueueInternal()
        }
    }
    
    /**
     * Removes a task from the queue (if it has not been started yet).
     */
    suspend fun removeFromQueue(requestId: String): Boolean = queueMutex.withLock {
        val removed = pendingQueue.remove(requestId)
        if (removed) {
            saveQueueAsync()
            Log.d(TAG, "Task removed from queue: $requestId")
        }
        return@withLock removed
    }
    
    /**
     * Checks whether a task is pending or active.
     */
    fun isTaskQueued(requestId: String): Boolean {
        return pendingQueue.contains(requestId)
    }
    
    fun isTaskActive(requestId: String): Boolean {
        return activeTasks.contains(requestId)
    }
    
    fun isTaskPendingOrActive(requestId: String): Boolean {
        return isTaskQueued(requestId) || isTaskActive(requestId)
    }
    
    /**
     * Returns queue statistics.
     */
    fun getQueueStats(): QueueStats {
        return QueueStats(
            pendingCount = pendingQueue.size,
            activeCount = activeTasks.size,
            maxConcurrent = maxConcurrentTasks,
            maxQueueSize = maxQueueSize
        )
    }
    
    /**
     * Sets the maximum number of concurrent tasks.
     */
    suspend fun setMaxConcurrentTasks(count: Int) {
        if (count < 1) {
            throw IllegalArgumentException("maxConcurrentTasks must be at least 1")
        }
        maxConcurrentTasks = count
        
        // If the limit was increased, try to start additional tasks
        queueMutex.withLock {
            processQueueInternal()
        }
    }
    
    /**
     * Sets the maximum queue size.
     */
    fun setMaxQueueSize(size: Int) {
        if (size < 1) {
            throw IllegalArgumentException("maxQueueSize must be at least 1")
        }
        maxQueueSize = size
    }
    
    /**
     * Clears the entire queue and cancels active tasks.
     */
    suspend fun clearAll(): Int = queueMutex.withLock {
        val totalCount = pendingQueue.size + activeTasks.size
        
        // Cancel active tasks in WorkManager
        for (requestId in activeTasks) {
            workManager.cancelRequest(requestId)
        }
        
        pendingQueue.clear()
        activeTasks.clear()
        saveQueueAsync()
        
        Log.d(TAG, "Cleared all tasks: $totalCount")
        return@withLock totalCount
    }
    
    /**
     * Forces queue processing.
     */
    suspend fun processQueue() = queueMutex.withLock {
        processQueueInternal()
    }
    
    // Internal queue processing method (must be called under mutex)
    private fun processQueueInternal() {
        while (activeTasks.size < maxConcurrentTasks && pendingQueue.isNotEmpty()) {
            val requestId = pendingQueue.poll() ?: break
            
            activeTasks.add(requestId)
            workManager.enqueueRequest(requestId)
            
            Log.d(TAG, "Started task: $requestId, active: ${activeTasks.size}, pending: ${pendingQueue.size}")
        }
        
        // Persist queue after changes
        saveQueueAsync()
    }
    
    // Persists the queue to disk asynchronously
    private fun saveQueueAsync() {
        scope.launch {
            try {
                val jsonArray = JSONArray()
                pendingQueue.forEach { jsonArray.put(it) }
                queueFile.writeText(jsonArray.toString())
            } catch (e: Exception) {
                Log.e(TAG, "Failed to save queue to disk", e)
            }
        }
    }
    
    // Restores the queue from disk
    private fun restoreQueue() {
        try {
            if (!queueFile.exists()) {
                Log.d(TAG, "No queue file found, starting fresh")
                return
            }
            
            val json = queueFile.readText()
            if (json.isBlank()) return
            
            val jsonArray = JSONArray(json)
            var restoredCount = 0
            
            for (i in 0 until jsonArray.length()) {
                val requestId = jsonArray.getString(i)
                pendingQueue.offer(requestId)
                restoredCount++
            }
            
            Log.d(TAG, "Restored $restoredCount tasks from disk")
            
            // Start processing of the restored queue
            scope.launch {
                queueMutex.withLock {
                    processQueueInternal()
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to restore queue from disk", e)
        }
    }
    
    /**
     * Synchronizes the queue state with the actual task state.
     * Called to clean up "stuck" tasks.
     */
    suspend fun syncQueueState(fileStorage: FileStorageDataSource) = queueMutex.withLock {
        val tasksToRemove = mutableListOf<String>()
        
        // Check active tasks
        for (requestId in activeTasks) {
            val workState = workManager.getWorkState(requestId, forceRefresh = true)
            when (workState) {
                WorkManagerDataSource.WorkStateResult.SUCCEEDED,
                WorkManagerDataSource.WorkStateResult.FAILED -> {
                    tasksToRemove.add(requestId)
                }
                WorkManagerDataSource.WorkStateResult.NOT_FOUND -> {
                    // Task is lost, re-enqueue
                    Log.w(TAG, "Task lost in WorkManager, re-enqueueing: $requestId")
                    workManager.enqueueRequest(requestId)
                }
                WorkManagerDataSource.WorkStateResult.IN_PROGRESS -> {
                    // Everything OK
                }
            }
        }
        
        // Remove completed from active
        tasksToRemove.forEach { activeTasks.remove(it) }
        
        // Check pending tasks – request files may have been deleted
        val pendingToRemove = mutableListOf<String>()
        for (requestId in pendingQueue) {
            if (!fileStorage.taskExists(requestId)) {
                pendingToRemove.add(requestId)
                Log.w(TAG, "Request file not found for pending task, removing: $requestId")
            }
        }
        pendingToRemove.forEach { pendingQueue.remove(it) }
        
        if (tasksToRemove.isNotEmpty() || pendingToRemove.isNotEmpty()) {
            saveQueueAsync()
            processQueueInternal()
        }
        
        Log.d(TAG, "Queue synced: removed ${tasksToRemove.size} completed, ${pendingToRemove.size} orphaned")
    }
    
    data class QueueStats(
        val pendingCount: Int,
        val activeCount: Int,
        val maxConcurrent: Int,
        val maxQueueSize: Int
    ) {
        fun toMap(): Map<String, Any> = mapOf(
            "pendingCount" to pendingCount,
            "activeCount" to activeCount,
            "maxConcurrent" to maxConcurrent,
            "maxQueueSize" to maxQueueSize
        )
    }
}

File: ios/Classes/Data/DataSource/TaskQueueManager.swift
import Foundation

/// Task queue manager for controlling the number of concurrent requests.
///
/// Solves the problem of hanging when registering a large number of tasks in URLSession.
/// Instead of starting all tasks at once, TaskQueueManager keeps them in its own queue
/// and starts them in batches.
actor TaskQueueManager {
    
    // MARK: - Singleton
    
    static let shared = TaskQueueManager()
    
    // MARK: - Constants
    
    private let queueFileName = "pending_queue.json"
    private let defaultMaxConcurrentTasks = 30
    private let defaultMaxQueueSize = 10000
    
    // MARK: - Properties
    
    /// Queue of pending tasks (only requestId, data is on disk)
    private var pendingQueue: [String] = []
    
    /// Set of active tasks
    private var activeTasks: Set<String> = []
    
    /// Callback for starting a task (set by the repository)
    private var executeTaskCallback: ((String) async -> Void)?
    
    /// Maximum number of concurrent tasks
    private(set) var maxConcurrentTasks: Int
    
    /// Maximum queue size
    private(set) var maxQueueSize: Int
    
    // MARK: - Initialization
    
    private init() {
        self.maxConcurrentTasks = defaultMaxConcurrentTasks
        self.maxQueueSize = defaultMaxQueueSize
        
        // Restore queue on initialization (synchronously)
        if let restored = Self.loadQueueFromDisk() {
            self.pendingQueue = restored
            print("[TaskQueueManager] Restored \(restored.count) tasks from disk")
        }
    }
    
    // MARK: - Public Methods
    
    /// Sets callback for executing a task
    func setExecuteCallback(_ callback: @escaping (String) async -> Void) async {
        self.executeTaskCallback = callback
        print("[TaskQueueManager] Execute callback set, processing pending queue...")
        // Process any tasks that were enqueued before callback was set
        await processQueueInternal()
    }
    
    /// Adds a task to the queue.
    /// - Parameter requestId: task ID
    /// - Returns: true if the task was added, false if the queue is full
    func enqueue(_ requestId: String) async -> Bool {
        // Check queue size limit
        guard pendingQueue.count < maxQueueSize else {
            print("[TaskQueueManager] Queue is full (\(maxQueueSize)), rejecting task: \(requestId)")
            return false
        }
        
        // Check that this task has not already been added
        guard !pendingQueue.contains(requestId) && !activeTasks.contains(requestId) else {
            print("[TaskQueueManager] Task already in queue or active: \(requestId)")
            return true
        }
        
        pendingQueue.append(requestId)
        saveQueueToDisk()
        
        print("[TaskQueueManager] Task enqueued: \(requestId), queue size: \(pendingQueue.count), active: \(activeTasks.count)")
        
        // Try to start tasks
        await processQueueInternal()
        
        return true
    }
    
    /// Called when a task is completed (success or error)
    func onTaskCompleted(_ requestId: String) async {
        if activeTasks.remove(requestId) != nil {
            print("[TaskQueueManager] Task completed: \(requestId), active: \(activeTasks.count)")
            await processQueueInternal()
        }
    }
    
    /// Removes a task from the queue (if it has not been started yet)
    func removeFromQueue(_ requestId: String) -> Bool {
        if let index = pendingQueue.firstIndex(of: requestId) {
            pendingQueue.remove(at: index)
            saveQueueToDisk()
            print("[TaskQueueManager] Task removed from queue: \(requestId)")
            return true
        }
        return false
    }
    
    /// Checks whether a task is in the queue
    func isTaskQueued(_ requestId: String) -> Bool {
        return pendingQueue.contains(requestId)
    }
    
    /// Checks whether a task is active
    func isTaskActive(_ requestId: String) -> Bool {
        return activeTasks.contains(requestId)
    }
    
    /// Checks whether a task is pending or active
    func isTaskPendingOrActive(_ requestId: String) -> Bool {
        return isTaskQueued(requestId) || isTaskActive(requestId)
    }
    
    /// Returns queue statistics
    func getQueueStats() -> QueueStats {
        return QueueStats(
            pendingCount: pendingQueue.count,
            activeCount: activeTasks.count,
            maxConcurrent: maxConcurrentTasks,
            maxQueueSize: maxQueueSize
        )
    }
    
    /// Sets the maximum number of concurrent tasks
    func setMaxConcurrentTasks(_ count: Int) async {
        guard count >= 1 else {
            print("[TaskQueueManager] maxConcurrentTasks must be at least 1")
            return
        }
        maxConcurrentTasks = count
        
        // If the limit was increased, try to start additional tasks
        await processQueueInternal()
    }
    
    /// Sets the maximum queue size
    func setMaxQueueSize(_ size: Int) {
        guard size >= 1 else {
            print("[TaskQueueManager] maxQueueSize must be at least 1")
            return
        }
        maxQueueSize = size
    }
    
    /// Clears the entire queue
    func clearAll() -> Int {
        let totalCount = pendingQueue.count + activeTasks.count
        
        pendingQueue.removeAll()
        activeTasks.removeAll()
        saveQueueToDisk()
        
        print("[TaskQueueManager] Cleared all tasks: \(totalCount)")
        return totalCount
    }
    
    /// Forces queue processing
    func processQueue() async {
        await processQueueInternal()
    }
    
    // MARK: - Private Methods
    
    private func processQueueInternal() async {
        // Do not process the queue if callback is not set yet
        guard executeTaskCallback != nil else {
            print("[TaskQueueManager] Cannot process queue: executeTaskCallback is not set")
            return
        }
        
        while activeTasks.count < maxConcurrentTasks && !pendingQueue.isEmpty {
            let requestId = pendingQueue.removeFirst()
            activeTasks.insert(requestId)
            
            print("[TaskQueueManager] Started task: \(requestId), active: \(activeTasks.count), pending: \(pendingQueue.count)")
            
            // Start the task via callback
            if let callback = executeTaskCallback {
                Task {
                    await callback(requestId)
                }
            }
        }
        
        // Save queue after changes
        saveQueueToDisk()
    }
    
    // MARK: - Persistence
    
    private var queueFileURL: URL {
        let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let dirURL = documentsURL.appendingPathComponent("background_http_client", isDirectory: true)
        
        // Create directory if it does not exist
        try? FileManager.default.createDirectory(at: dirURL, withIntermediateDirectories: true)
        
        return dirURL.appendingPathComponent(queueFileName)
    }
    
    private func saveQueueToDisk() {
        do {
            let data = try JSONEncoder().encode(pendingQueue)
            try data.write(to: queueFileURL)
        } catch {
            print("[TaskQueueManager] Failed to save queue to disk: \(error)")
        }
    }
    
    private static func loadQueueFromDisk() -> [String]? {
        let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let fileURL = documentsURL
            .appendingPathComponent("background_http_client", isDirectory: true)
            .appendingPathComponent("pending_queue.json")
        
        guard FileManager.default.fileExists(atPath: fileURL.path) else {
            return nil
        }
        
        do {
            let data = try Data(contentsOf: fileURL)
            return try JSONDecoder().decode([String].self, from: data)
        } catch {
            print("[TaskQueueManager] Failed to load queue from disk: \(error)")
            return nil
        }
    }
}

// MARK: - Models

struct QueueStats {
    let pendingCount: Int
    let activeCount: Int
    let maxConcurrent: Int
    let maxQueueSize: Int
    
    func toDict() -> [String: Any] {
        return [
            "pendingCount": pendingCount,
            "activeCount": activeCount,
            "maxConcurrent": maxConcurrent,
            "maxQueueSize": maxQueueSize
        ]
    }
}

